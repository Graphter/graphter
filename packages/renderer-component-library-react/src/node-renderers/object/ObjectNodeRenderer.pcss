.objectNodeRenderer{
    display: inline-block;
    padding: 20px;
}
.defaultWrapper{

}
.field-name{

}
.description{
    font-size: 0.7em;
    margin-top: 0;
}