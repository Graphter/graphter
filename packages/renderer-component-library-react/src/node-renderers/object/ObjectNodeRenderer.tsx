import React from "react";
import { NodeConfig, NodeRendererProps } from "@graphter/core";
import { nodeRendererStore } from "@graphter/renderer-react";
import { createDefault, useNodeData } from "@graphter/renderer-react";
import s from './ObjectNodeRenderer.pcss'

export default function ObjectNodeRenderer(
  {
    config,
    originalNodeData,
    path,
    committed,
    ErrorDisplayComponent
  }: NodeRendererProps
) {
  if (!originalNodeData) originalNodeData = createDefault(config, {})
  useNodeData(path, config, originalNodeData, committed)
  return (
    <div className={s.objectNodeRenderer}>
      {config.children && config.children.map((childConfig, i) => {
        const childRendererRegistration = nodeRendererStore.get(childConfig.type)
        if (!childRendererRegistration) return null
        const ChildTypeRenderer = childRendererRegistration.renderer
        return (
          <DefaultPropertyWrapper config={childConfig} key={childConfig.id}>
            <ChildTypeRenderer
              committed={committed}
              config={childConfig}
              path={[ ...path, childConfig.id ]}
              originalNodeData={originalNodeData[childConfig.id]}
              options={childRendererRegistration.options}
              ErrorDisplayComponent={ErrorDisplayComponent} />
          </DefaultPropertyWrapper>
        )
      })}
    </div>
  );
}

interface DefaultPropertyWrapperProps {
  config: NodeConfig
  children: any
}

function DefaultPropertyWrapper(
  {
    config,
    children
  }: DefaultPropertyWrapperProps
) {
  return (

    <div className={s.defaultWrapper}>
      <label htmlFor={config.id}>{config.name}</label>
      {config.description && <p className={s.description}>{config.description}</p>}
      {children}
    </div>
  )
}