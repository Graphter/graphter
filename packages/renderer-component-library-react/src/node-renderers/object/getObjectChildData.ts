import { GetChildDataFn } from "@graphter/core";
import { nodeRendererStore } from "@graphter/renderer-react";

export const getObjectChildData: GetChildDataFn = (config, path, getNodeValue) => {
  if (!config.children?.length) throw new Error(`${config.type} type '${config.id}' has no children configured. At least one is required.`)

  const childValues = config.children.map(childConfig => {
    const childPath = [ ...path, childConfig.id ]
    const childRenderer = nodeRendererStore.get(childConfig.type)
    const nodeData = childRenderer.getChildData ?
      childRenderer.getChildData(childConfig, childPath, getNodeValue) :
      getNodeValue(childPath)
    return { key: childConfig.id, value: nodeData }
  })
  const reFormedObject = childValues.reduce<{ [key: string]: any }>((a, c) => {
    a[c.key] = c.value
    return a
  }, {})
  console.info(`Reformed ${childValues.length} child values for node '${config.id}'`, reFormedObject)
  return reFormedObject
}