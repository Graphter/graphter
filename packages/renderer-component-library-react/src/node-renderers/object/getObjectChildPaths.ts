import { GetChildPathsFn } from "@graphter/core";
import { nodeRendererStore } from "@graphter/renderer-react";

export const getObjectChildPaths: GetChildPathsFn = (config, path, getNodeValue) => {
  if(!config.children?.length) throw new Error(`${config.type} type '${config.id}' has no children configured. At least one is required.`)
  const results = config.children.map(childConfig => {
    const childPath = [ ...path, childConfig.id ]
    const childRenderer = nodeRendererStore.get(childConfig.type)
    if(!childRenderer.getChildPaths) return [ childPath ]
    return [ childPath, ...childRenderer.getChildPaths(childConfig, childPath, getNodeValue) ]
  })
  const paths = results.flat()
  console.info(`Found ${paths.length} paths for tree starting from '${path.join('/')}'`, paths)
  return paths
}