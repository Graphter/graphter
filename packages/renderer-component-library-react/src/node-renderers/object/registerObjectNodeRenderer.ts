import ObjectNodeRenderer from "./ObjectNodeRenderer";
import { NodeRendererRegistration } from "@graphter/core";
import { getObjectChildData } from "./getObjectChildData";
import { getObjectChildPaths } from "./getObjectChildPaths";
import { initialiseObjectData } from "./initialiseObjectData";
import { getObjectChildConfig } from "./getObjectChildConfig";

export interface ObjectNodeRendererOptions {
  type: string
}

export function registerObjectNodeRenderer(options?: ObjectNodeRendererOptions): NodeRendererRegistration {
  const type = options?.type || 'object'
  return {
    type: type,
    name: 'Object',
    description: 'Manage complex nested structures',
    initialiseData: initialiseObjectData,
    getChildConfig: getObjectChildConfig,
    getChildData: getObjectChildData,
    getChildPaths: getObjectChildPaths,
    Renderer: ObjectNodeRenderer,
    options
  }
}