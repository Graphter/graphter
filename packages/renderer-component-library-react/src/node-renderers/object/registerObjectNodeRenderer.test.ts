import { getObjectChildData } from "./getObjectChildData";
import { getObjectChildPaths } from "./getObjectChildPaths";
import ObjectNodeRenderer from "./ObjectNodeRenderer";
import { registerObjectNodeRenderer } from "./registerObjectNodeRenderer";

describe('registerObjectNodeRenderer()', () => {
  it('should return the correct registration', async () => {
    const result = registerObjectNodeRenderer()
    expect(result).toMatchSnapshot()
    expect(result.getChildData).toBe(getObjectChildData)
    expect(result.getChildPaths).toBe(getObjectChildPaths)
    expect(result.Renderer).toBe(ObjectNodeRenderer)
  })
  it('should override the type if supplied in options', async () => {
    const result = registerObjectNodeRenderer({ type: 'custom-object' })
    expect(result.type).toBe('custom-object')
  })
})