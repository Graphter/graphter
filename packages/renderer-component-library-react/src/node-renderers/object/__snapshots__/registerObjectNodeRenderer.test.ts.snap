// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`registerObjectNodeRenderer() should return the correct registration 1`] = `
Object {
  "Renderer": [Function],
  "description": "Manage complex nested structures",
  "getChildData": [Function],
  "getChildPaths": [Function],
  "initialiseData": [Function],
  "name": "Object",
  "options": undefined,
  "type": "object",
}
`;
